### PR TITLE
Add shader pointer parameter tests, including for `unrestricted_pointer_parameter` support

### PR DESCRIPTION
**Pull request by ben-clayton**
New execution tests for pointer parameters, and updated validation tests to handle the `unrestricted_pointer_parameters` WGSL feature.

Tested with and without `unrestricted_pointer_parameters` being supported.

Issue: #3056

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Review by dneto0 — APPROVED** [review]
Thanks! These are very good.

I noted a followup to add matching type cases for 3 subtests.

File: src/webgpu/gpu_test.ts [hunk 1]
```diff
@@ -459,21 +459,25 @@ export class GPUTestBase extends Fixture<GPUTestSubcaseBatchState> {
   }
 
   /** Skips this test case if the `langFeature` is *not* supported. */
-  requireLanguageFeatureOrSkipTestCase(langFeature: WGSLLanguageFeature) {
-    const lf = getGPU(this.rec).wgslLanguageFeatures;
-    if (lf === undefined || !lf.has(langFeature)) {
+  skipIfLanguageFeatureNotSupported(langFeature: WGSLLanguageFeature) {
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/gpu_test.ts:462]: Ack. :)

File: src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+`;
+  }
+}
+
+function valuesForType(kind: 'vec4i' | 'array' | 'struct') {
+  switch (kind) {
+    case 'vec4i':
+      return new Uint32Array([1, 2, 3, 4]);
+    case 'array':
+      return new Float32Array([1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13]);
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts:38]: I can count to 12. I promise.

File: src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+fn main() {
+  f0(&input);
+}
+`,
+    }[t.params.address_space];
+
+    let call_chain = '';
+    for (let i = 0; i < t.params.call_indirection; i++) {
+      call_chain += `
+fn f${i}(p : ptr<${t.params.address_space}, T>) {
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts:145]: It is. It's also annoying that `uniform` and others don't allow `read` access, as it means we have to specialize the generation. The language is inconsistent here, and we should probably fix it.

File: src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+
+    // prettier-ignore
+    const input = new Uint32Array([
+      /* S.b.a */ 5, 6, 7, 8,
+      /* S.b.b */ 9, 10, 11, 12,
+    ]);
+
+    // prettier-ignore
+    const expected = new Uint32Array([
+      /* S.a   */ 0, 0, 0, 0,
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts:538]: ack

File: src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts]: Done

File: src/webgpu/shader/validation/functions/restrictions.spec.ts [hunk 5]
```diff
@@ -426,15 +440,39 @@ const kFunctionParamValueCases: Record<string, ParamValueCase> = {
...
+    value: `&g_struct_with_array.a[2]`,
+    matches: ['ptr4'],
+    needsUnrestrictedPointerParameters: true,
+  },
+  ptr9: {
+    value: `&ro_constructible.b`,
+    matches: ['ptr5', 'ptr6'],
+    needsUnrestrictedPointerParameters: true,
+  },
+  ptr10: { value: `&rw_constructible`, matches: [], needsUnrestrictedPointerParameters: true },
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/shader/validation/functions/restrictions.spec.ts]: Done

File: src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dneto0) [reply src/webgpu/shader/execution/expression/call/user/ptr_params.spec.ts:145]: It means more test cases is in more places.  It's a reasonable choice, IMHO.

**Review by dneto0 — APPROVED** [review]
Approved new changes too
thanks!

**Pull request merged**